### DISCUSSION TITLE
Progressive Web Apps

### DISCUSSION THREAD

**Discussion by yurimatheus** [frontendbr/forum#34] [Perguntas e Respostas]
Estava dando uma lida em alguns artigos para me atualizar, e esse aqui (http://arquiteturadeinformacao.com/mobile/o-que-sao-progressive-web-apps/) do blog Arquitetura da Informação me atraiu a atenção.

Fiz uns testes em dois links que se encontram no artigo (Chrome Dev Summit e o app de shopping do @sergiolopes) que ficam perfeitos no Android. Caso queiram visualizar o app, acessem: https://sergiolopes.github.io/shopping/ e 'adicione à tela inicial'.

Alguém de vocês já viu? O que acham? Será que é mais fácil fazer apps assim do que com IONIC?

**Comment by raphaklaus** [comment]
Boa @yurimatheus! Particularmente acho que esse é o caminho certo para a web.
Eu usei o Ionic Framework durante 1 ano e pude ver a evolução da ferramenta, mas sempre achei meio estranho utilizar uma **webview** dentro de um app só pra poder usar os recursos do celular (geo, camera, etc). Com a evolução dos browsers acho que essa técnica de "embedar" vai começar a ficar menos popular por motivos óbvios, até porque é muito burocrática... Logo web puro, é mais facil. Por quê?
- Não precisa esperar aprovação da loja.
- Não precisa gerar um apk ou semelhante.
- Não precisa gerar chaves.
- Dependendo do browser, não precisa instalar vários plugins para lidar com operações nativas.
- Não obriga usuário a baixar app.

Outra coisa também que já pensei é no interesse das lojas (AppStore, GooglePlay e Windows Store) em manter esse tipo de negócio ativo, pois é lucrativo para eles. Quando comprei a minha licença, Android tava U$ 25,00, Apple U$ 99,00 e Windows U$ 19,00... Vamos acompanhar essa novela... :stuck_out_tongue:

**Comment by willianjusten** [comment]
Esse ano teremos 3 coisas em grande ascenção na minha opinião:
- Service Workers
- Progressive Web App
- Streaming

E indo na linha de progressive web app, tem dois artigos incríveis que eu aconselho bastante, que são:
- [Progressive Web Apps by Google](https://developers.google.com/web/progressive-web-apps)
- [Getting started with Progressive Web Apps](https://addyosmani.com/blog/getting-started-with-progressive-web-apps/)

Mas eu não acho que esse tipo de coisa vá brigar lado a lado com apps não. 

1 - Praticamente ninguém adiciona site na tela inicial.
2 - Usuários leigos ainda não estão acostumados a esse tipo de uso.

Eu acho que sim, esse tipo de aplicação tem futuro e vai vingar para alguns nichos, mas acho que não necessariamente para todos.

**Comment by fernandofleury** [comment]
Web App é foda, mas a gente ainda enfrenta o que as empresas tomam como relevante, e o que agrega valor. (a ideia do aplicativo _existir_ numa store da vida, pra muita gente é um valor absurdamente grande)

**Comment by felipefialho** [comment]
Comecei a me interessar pelo assunto, vendo uma puta palestra do @sergiolopes sobre o assunto e concordo totalmente. 

Se vocês observarem, a grande maioria dos aplicativos, especialmente os vinculados a sites ou serviços online são desnecessários e poderiam ser facilmente resolvidos usando um web app.

O exemplo que ele deu do shopping é real, é quase inacreditável você precisar baixar um app pra pagar um serviço de algo que você vai usar uma vez na vida outra na morte. Eu sou muito chato quanto aos aplicativos instalados no celular e prefiro usar os sites quando tenho a opção.

Mas o @willianjusten falou uma coisa válida. Eu também acho muito dificil mudar a mente do usuário para ele começar a instalar web app no celular. Eu mesmo nunca fiz isso. Então o engajamento acaba sendo bem menor.

**Comment by felquis** [comment]
Na startup que trabalho estamos trabalhando com Web App, como temos apenas app para Android, Windows Phone e iOS são atendidos 100% via Web App.

No iOS temos um estimulo "tutorial" que é exibido em um determinado momento depois que o usuário está usando o site, para que ele adicione a tela inicial. É exibido apenas uma vez no device, via localStorage. e via userAgent sniffing

No Android usamos Service Workers e funciona bem, a navegação mesmo mudando de URL é muito rápida. E o Chrome automaticamente exibe para o usuário um prompt, perguntando se o usuário quer adicionar a homescreen. Ainda não usamos Push Notifications, porém já está engatilhado.

No Windows Phone também há como fixar o site na homescreen, porém ainda não investimos em um tutorial parecido com o que usamos no iOS.

Fizemos apenas uma versão mobile, agora estou trabalhando em uma versão desktop em cima do mesmo código... O objetivo do projeto não é ser um "site", e sim o essencial para satisfazer os usuários.. O que facilita muita coisa. No desktop nosso querido visionário diz ser "O painel de controle do usuário", mais espaço na tela, mais componentes.

> Será que é mais fácil fazer apps assim do que com IONIC?

Com certeza é mais `fácil` fazer Web Apps com as tecnologias nativas, com fluência em CSS, HTML e JavaScript é muito mais simples alcançar o essencial, se vc começa com 100kb de vendors, vc já saiu do essencial. Para mim isso já é um Web App -> http://codepen.io/felquis/pen/zrerjX  full http://s.codepen.io/felquis/debug/zrerjX

Tbm trabalhei com Ionic, ele é um meio. Porém existe N formas de resolver os problemas, e todas as formas estão corretas, desde que o usuário complete seu objetivo no site. O stack aqui essencialmente é API Rest, o server que entrega o Front por enquanto é o Sails. Templates pré-compilados com Jade, RequireJS + Almond.. dependencias do Front são: 
- autolinker
- cache-polyfill.js
- canvas-to-blob
- cookie-monster
- crossroads + signals
- es6-promises polyfill
- image-resize com canvas
- jade-runtime
- swiper
- microjs translate

Com jade, estamos caminhando para isomorphic application. Porém não será 100%, será apenas o essencial para buscadores. E para um instant load, enquanto o restante da aplicação se molda enquanto é carregada. Isso é de certa forma uma tarefa bem simples.

Todos estão se voltando a Web "Apps", mas na verdade só estamos alcançando a essência da Web como ela sempre deveria ter sido, como a sonhamos

Vida longa aos Web Apps, pois eu já enterrei essas "webviews".

> Eu também acho muito dificil mudar a mente do usuário para ele começar a instalar web app no celular.

No Windows Phone, multitask é quase impossível, uso um Lumia 640. Porém no Internet Explorer, multitask é muito mais intuitivo. Acho que não foi de propósito deles, eu que sou mais acostumado com Android mesmo... Porém instalar o Web App será uma tarefa do navegador, o Chrome está fazendo isso, já exibe um prompt e tal, nas outras plataformas com um pouco de esforço podemos influenciar o usuário com carinho a adicionar o web app a home screen, pois não da pra esperar a mesma boa vontade do Safari por exemplo, do Firefox e do Opera até da para esperar essa boa vontade. Mas o Safari duvido muito que exibiriam um prompt automático. Minha família usa Web Apps adicionados a home screens, e nem percebem. Com muita influência minha, porém lembro que uma vez sai de casa e deixei o app Android do Facebook instalado para minha mãe, 1 semana depois quando voltei, ela estava usando o site do Facebook, na home screen, e só Deus sabe como ele foi parar la, ela usa o Facebook na Web até hoje. Minha Mãe, que é um usuário bem noob com tecnologia. Então isso me faz acreditar que há muita esperança.

**Comment by vitor-mendrone-deel** [comment]
Concordo com o @felquis (que alias mandou bem demais na postagem, bem detalhada) o trabalho de "educar" os usuários a adicionarem o webapp a home dependerá exclusivamente dos navegadores. Pela simplicidade é bem possível que a Microsoft, a Apple e a Mozilla invistam em soluções parecidas com a adotada no Chrome e convide o usuário a adicionar a aplicação a home. 

Vejo os PWAs como o futuro das aplicações web mobile. Como disse o @sergiolopes, ninguém quer fazer o download de um app pra pagar o estacionamento ou pra automatizar o cardápio em um restaurante.

**Comment by thierryrene** [comment]
Caras, recentemente vi uma enxurrada de artigos sobre PWA e AMP, e será que de fato não é esse o melhor caminho para aplicações simples? por exemplo, em um dos artigos, era mencionado que como um PWA tem algumas integrações com o Android (e novas integrações estão a caminho), não perdemos tanto tempo em uma solução simples focada na plataforma. Sem contar os service workers que ajudam muito.

Bom, alguém tem algo feito com ele pra mostrar sem ser o site do WSJ (washington post)?

edit:

Link para os artigos que li.

[Why “Progressive Web Apps vs. native” is the wrong question to ask](https://medium.com/dev-channel/why-progressive-web-apps-vs-native-is-the-wrong-question-to-ask-fb8555addcbb?goal=0_df65b6d7c8-2467c78558-42764681&mc_cid=2467c78558&mc_eid=89382533bf#.dip5hq18l)

[WTF are Progressive Web Apps?](http://digiday.com/platforms/wtf-progressive-web-apps/)

[Progressive web apps – let’s not repeat the errors from the beginning of responsive web design](https://justmarkup.com/log/2016/06/progressive-web-apps-lets-not-repeat-the-errors-from-the-beginning-of-responsive-web-design/)

Valeus.

**Comment by williamcosta** [comment]
Pessoal tem um [podcast](http://hipsters.tech/progressive-web-apps-hipsters-03/) muito bom falando sobre WebApps. Quem puder ouvir recomendo.

**Comment by thierryrene** [comment]
Vou dar uma olhada valeu @williamcosta

**Comment by yurimatheus** [comment]
@williamcosta o podcast abre bem a mente, me ajudou bastante. O exemplo do git lá do @sergiolopes me deu uma ideia de estrutura do webapp.

@thierryrene os links foram sensacionais cara.

**Comment by yurimatheus** [comment]
**Desculpem... estava pelo celular e cliquei errado!**

. . . caso desejem dá uma estudada em PWA com Angular segue link : https://github.com/angular/mobile-toolkit

**Comment by marcoblos** [comment]
@williamcosta tenho acompanhado sempre o [hipsters.tech](http://hipsters.tech/), na minha opinião, o melhor podcast atual, muito direto e entrevistas sempre interessantes.

**Comment by williamcosta** [comment]
Sim @marcoblos estou acompanhando também é um dos melhores em Português e sobre tecnologia da atualidade.

**Comment by saitodisse** [comment]
To curtindo tbm o hipsters. O Dev na Estrada tbm é bacana. Hehehe. O podia
ter uma issue sobre podcasts.

Em 26 de set de 2016 14:45, "William Costa" notifications@github.com
escreveu:

**Comment by felquis** [comment]
Fiz um video de 20 minutos onde passo por todas as principais partes de um PWA, acho que vocês podem gostar [A Era dos Progressive Web Apps](https://www.youtube.com/watch?v=AOWtO51pmHo)

Abraço.

**Comment by ramonsenadev** [comment]
Nessa semana eu estava lendo uma série de postagens do Addy Osmani no medium sobre PWA com React. Estou gostando bastante, elas realmente são bem legais e eu recomendo para quem ainda não leu. Logo mais ele postará a parte 5. Enfim, segue o link da parte 1: https://medium.com/@addyosmani/progressive-web-apps-with-react-js-part-i-introduction-50679aef2b12#.izyw0e9jy

**Comment by ngoctrai1997** [comment]
esta página ajuda você a entender mais sobre o pwa, você pode visitá-lo : [https://tapita.io/](url)
